subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/recomp/AbstractRecompilationSpecProvider.java

```diff
@@ -16,38 +16,140 @@
+        // In the same situation with NATIVE_HEADER_OUTPUT, javac just NPEs.  Don't bother.
+        cleanedAnyOutput |= deleteStaleFilesIn(resourcesToDelete.get(GeneratedResource.Location.NATIVE_HEADER_OUTPUT), spec.getCompileOptions().getHeaderOutputDirectory());
+
+        return cleanedAnyOutput;
+    }
+
+    private Iterable<File> narrowDownSourcesToCompile(FileTree sourceTree, PatternSet sourceToCompile) {
+        return sourceTree.matching(sourceToCompile);
+    }
+
+    private static Map<GeneratedResource.Location, PatternSet> prepareResourcePatterns(Collection<GeneratedResource> staleResources, FileOperations fileOperations) {
+        Map<GeneratedResource.Location, PatternSet> resourcesByLocation = new EnumMap<>(GeneratedResource.Location.class);
+        for (GeneratedResource.Location location : GeneratedResource.Location.values()) {
+            resourcesByLocation.put(location, fileOperations.patternSet());
+        }
+        for (GeneratedResource resource : staleResources) {
+            resourcesByLocation.get(resource.getLocation()).include(resource.getPath());
+        }
+        return resourcesByLocation;
+    }
+
+    private void processOtherChanges(CurrentCompilation current, PreviousCompilation previous, RecompilationSpec spec, SourceFileClassNameConverter sourceFileClassNameConverter) {
+        if (spec.isFullRebuildNeeded()) {
+            return;
+        }
+        SourceFileChangeProcessor sourceFileChangeProcessor = new SourceFileChangeProcessor(previous);
+        for (FileChange fileChange : sourceChanges) {
+            if (spec.isFullRebuildNeeded()) {
+                return;
+            }
+            if (fileChange.getFileType() != FileType.FILE) {
+                continue;
+            }
+
+            String relativeFilePath = fileChange.getNormalizedPath();
+            Set<String> changedClasses = sourceFileClassNameConverter.getClassNames(relativeFilePath);
+            if (changedClasses.isEmpty() && !isIncrementalOnResourceChanges(current)) {
+                spec.setFullRebuildCause(rebuildClauseForChangedNonSourceFile(fileChange));
+            }
+            sourceFileChangeProcessor.processChange(changedClasses, spec);
+        }
+    }
+
+    protected abstract boolean isIncrementalOnResourceChanges(CurrentCompilation currentCompilation);
+
+    private void prepareFilePatterns(Collection<String> staleClasses, PatternSet filesToDelete, PatternSet sourceToCompile, SourceFileClassNameConverter sourceFileClassNameConverter) {
+        for (String staleClass : staleClasses) {
+            for (String sourcePath : sourceFileClassNameConverter.getRelativeSourcePaths(staleClass)) {
+                filesToDelete.include(sourcePath);
+                sourceToCompile.include(sourcePath);
+            }
+            filesToDelete.include(staleClass.replaceAll("\\.", "/").concat(".class"));
```

Comment:
I think if we don't have `compilerApiData`, then we need to add the old post-fixes here as well, right? I don't see how this would work without having the mappings file from the Java compiler plugin.

Reply:
We run the same tests for CLI and daemon compiler and they all pass with this. I'll have to check for missing coverage then.

Reply:
I am really suprised that the inner classes are removed when we remove them from the source file without having the `$*.class` extra include. Do they reference the class and that is why we remove them nontheless?

Reply:
I played a little bit around and it seems like we find inner and anonymous classes as dependents of the actual classes. That is why this works I suppose.